libcore/src/geometry/helper/CorrectGeometry.cpp

```diff
@@ -301,238 +301,332 @@ void CorrectInputGeometry(Building & building)
+        });
+
+    std::for_each(
+        std::begin(removedWalls),
+        std::end(removedWalls),
+        [trainID, &building, &subroom](const Wall & wall) {
+            building.AddTrainWallRemoved(trainID, wall);
+            subroom.RemoveWall(wall);
+        });
+
+    std::for_each(
+        std::begin(trainDoors),
+        std::end(trainDoors),
+        [trainID, &building, &room, &subroom](const Transition & door) {
+            auto trainDoor = new Transition(door);
+            building.AddTrainDoorAdded(trainID, door);
+            room->AddTransitionID(trainDoor->GetUniqueID()); // danger area
+            subroom.AddTransition(trainDoor);                // danger area
+            building.AddTransition(trainDoor);               // danger area
+        });
 
-        std::for_each(
-            std::begin(addedWalls),
-            std::end(addedWalls),
-            [trainID, &building, &subroom](const Wall & wall) {
-                building.AddTrainWallAdded(trainID, wall);
-                subroom.AddWall(wall);
-            });
-        std::for_each(
-            std::begin(removedWalls),
-            std::end(removedWalls),
-            [trainID, &building, &subroom](const Wall & wall) {
-                building.AddTrainWallRemoved(trainID, wall);
-                subroom.RemoveWall(wall);
-            });
-        building.AddTrainDoorAdded(trainID, *trainDoor);
-    }
     subroom.Update();
 }
 
-std::tuple<std::vector<Wall>, std::vector<Wall>> SplitWall(
-    const std::pair<std::pair<Point, Wall>, std::pair<Point, Wall>> & wallDoorIntersectionPoints,
-    const std::vector<Wall> & trackWalls,
-    const Transition & door)
+std::tuple<std::vector<Wall>, std::vector<Wall>>
+SplitWalls(const std::vector<Wall> & trackWalls, const std::vector<Transition> & doors)
 {
+    std::vector<Wall> walls{trackWalls};
     std::vector<Wall> addedWalls;
     std::vector<Wall> removedWalls;
 
-    auto pw1 = wallDoorIntersectionPoints.first;
-    auto pw2 = wallDoorIntersectionPoints.second;
-    auto p1  = pw1.first;
-    auto w1  = pw1.second;
-    auto p2  = pw2.first;
-    auto w2  = pw2.second;
-
-    // case 1
-    Point P;
-    if(w1 == w2) {
-        double dist_pt1 = (w1.GetPoint1() - door.GetPoint1()).NormSquare();
-        double dist_pt2 = (w1.GetPoint1() - door.GetPoint2()).NormSquare();
-        Point A, B;
-
-        if(dist_pt1 < dist_pt2) {
-            A = door.GetPoint1();
-            B = door.GetPoint2();
-        } else {
-            A = door.GetPoint2();
-            B = door.GetPoint1();
+    for(auto const & door : doors) {
+        auto wallStartItr =
+            std::find_if(std::begin(walls), std::end(walls), [&door](const Wall & wall) {
+                return wall.IsInLineSegment(door.GetPoint1()) &&
+                       wall.GetPoint2() != door.GetPoint1();
+            });
+        if(wallStartItr == std::end(walls)) {
+            throw std::runtime_error("");
         }
```

Comment:
what happens if there is no door at a track wall? might be a valid scenario for short track walls. I think that should not lead to an error

Reply:
It is checked if all door points can be found on a wall. As `doors` comes from `ComputeTrainDoorCoordinates` hence it should never happen that a door point does not belong to any of the given track walls.